crates/subspace-service/src/domains/request_handler.rs

```diff
@@ -0,0 +1,268 @@
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+use domain_runtime_primitives::Balance;
+use futures::channel::oneshot;
+use futures::stream::StreamExt;
+use parity_scale_codec::{Decode, Encode};
+use sc_client_api::{BlockBackend, ProofProvider};
+use sc_domains::FPStorageKeyProvider;
+use sc_network::config::ProtocolId;
+use sc_network::request_responses::{IncomingRequest, OutgoingResponse};
+use sc_network::{NetworkBackend, PeerId};
+use sp_api::ProvideRuntimeApi;
+use sp_blockchain::HeaderBackend;
+use sp_domains::{DomainId, DomainsApi, ExecutionReceiptFor};
+use sp_domains_fraud_proof::storage_proof::{
+    BasicStorageProof, LastConfirmedDomainBlockReceiptProof,
+};
+use sp_domains_fraud_proof::FraudProofApi;
+use sp_runtime::codec;
+use sp_runtime::traits::{Block as BlockT, Header};
+use std::marker::PhantomData;
+use std::sync::Arc;
+use std::time::Duration;
+use tracing::{debug, error, trace};
+
+/// Generates a `RequestResponseProtocolConfig` for the state request protocol, refusing incoming
+/// requests.
+pub fn generate_protocol_config<
+    Hash: AsRef<[u8]>,
+    B: BlockT,
+    N: NetworkBackend<B, <B as BlockT>::Hash>,
+>(
+    _: &ProtocolId,
+    genesis_hash: Hash,
+    fork_id: Option<&str>,
+    inbound_queue: async_channel::Sender<IncomingRequest>,
+) -> N::RequestResponseProtocolConfig {
+    N::request_response_config(
+        generate_protocol_name(genesis_hash, fork_id).into(),
+        Vec::new(),
+        1024 * 1024,
+        16 * 1024 * 1024,
+        Duration::from_secs(40),
+        Some(inbound_queue),
+    )
+}
+
+/// Generate the state protocol name from the genesis hash and fork id.
+pub fn generate_protocol_name<Hash: AsRef<[u8]>>(
+    genesis_hash: Hash,
+    fork_id: Option<&str>,
+) -> String {
+    let genesis_hash = genesis_hash.as_ref();
+    if let Some(fork_id) = fork_id {
+        format!(
+            "/{}/{}/last-confirmed-block/1",
+            array_bytes::bytes2hex("", genesis_hash),
+            fork_id
+        )
+    } else {
+        format!(
+            "/{}/last-confirmed-block/1",
+            array_bytes::bytes2hex("", genesis_hash)
+        )
+    }
+}
+
+/// Request last confirmed domain block data from a peer.
+#[allow(clippy::derive_partial_eq_without_eq)]
+#[derive(Clone, PartialEq, Encode, Decode, Debug)]
+pub struct LastConfirmedBlockRequest {
+    pub domain_id: DomainId,
+}
+
+#[allow(clippy::derive_partial_eq_without_eq)]
+#[derive(Clone, PartialEq, Encode, Decode, Debug)]
+pub struct LastConfirmedBlockResponse<Block: BlockT, DomainHeader: Header> {
+    pub last_confirmed_block_data: Option<(
```

Comment:
It should not be `Option` because the `LatestConfirmedDomainExecutionReceipt` must be a non-empty value for a given domain id otherwise the storage proof generation/verification will fail.

When a domain just instantiated and haven't pass its first challenge period the `LatestConfirmedDomainExecutionReceipt` is storing the genesis receipt:
https://github.com/autonomys/subspace/blob/38d6b5be8209b9ad5cd08a788dff7675e621540f/crates/sp-domains/src/lib.rs#L655-L674

where the `consensus_block_hash` is the default value and the `consensus_block_number` is zero, we may need to set these fields to a concrete value for the domain sync to work.